File: lib/JFFNetwork/HttpBodyBuilder/JHttpBodyBuilder.m
#import "JHttpBodyBuilder.h"

@implementation JHttpBodyBuilder

+ (NSString *)tmpFileForUploadStreamWithDataForFilePath:(NSString *)dataFilePath
                                               boundary:(NSString *)boundary
                                                   name:(NSString *)name
                                               fileName:(NSString *)fileName
                                          dictWithParam:(NSDictionary *)dictWithParam
{
    NSString *filePath;
    
    @autoreleasepool {
        
        filePath = [NSString createUuid];
        filePath = [NSString cachesPathByAppendingPathComponent:filePath];
        const char *filePathPtr = [filePath cStringUsingEncoding:NSUTF8StringEncoding];
        
        FILE *file = fopen(filePathPtr, "w+");
        
        @autoreleasepool
        {
            {
                NSString *boundaryStr = [[NSString alloc] initWithFormat:@"--%@\r\n", boundary];
                NSData *boundaryData = [boundaryStr dataUsingEncoding:NSUTF8StringEncoding];
                fwrite([boundaryData bytes], 1, [boundaryData length], file);
            }
            //[result appendData:[[[NSString alloc] initWithFormat:@"--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
            
            {
                NSString *contentDisposition = [[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"; filename=\"%@\"\r\n", name, fileName];
                NSData *contentDispositionData = [contentDisposition dataUsingEncoding:NSUTF8StringEncoding];
                fwrite([contentDispositionData bytes], 1, [contentDispositionData length], file);
            }
            //[result appendData:[[[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"; filename=\"%@\"\r\n", parameter, fileName] dataUsingEncoding:NSUTF8StringEncoding]];
            
            {
                NSString *contentType = @"Content-Type: application/octet-stream\r\n\r\n";
                NSData *contentTypeData = [contentType dataUsingEncoding:NSUTF8StringEncoding];
                fwrite([contentTypeData bytes], 1, [contentTypeData length], file);
            }
            //[result appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
            
            {
                FILE *uploadDataFile = fopen([dataFilePath cStringUsingEncoding:NSUTF8StringEncoding],
                                        "r");
                
                uint8_t buffer[10*1024];
                const size_t bufferSize = sizeof(buffer)/sizeof(buffer[0]);
                
                size_t readBytes = 0;
                
                while ((readBytes = fread((void* __restrict)&buffer[0], 1, bufferSize, uploadDataFile)) != 0) {
                    
                    fwrite(buffer, 1, readBytes, file);
                }
                
                fclose(uploadDataFile);
            }
            //[result appendData:data];
            
            {
                NSString *boundaryStr = [[NSString alloc] initWithFormat:@"\r\n--%@\r\n", boundary];
                NSData *boundaryData = [boundaryStr dataUsingEncoding:NSUTF8StringEncoding];
                fwrite([boundaryData bytes], 1, [boundaryData length], file);
            }
            //[result appendData:[[[NSString alloc] initWithFormat:@"\r\n--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
        }
        
        {
            [dictWithParam enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
                
                @autoreleasepool
                {
                    {
                        NSString *boundaryStr = [[NSString alloc] initWithFormat:@"--%@\r\n", boundary];
                        NSData *boundaryData = [boundaryStr dataUsingEncoding:NSUTF8StringEncoding];
                        fwrite([boundaryData bytes], 1, [boundaryData length], file);
                    }
                    //[self appendData:[[[NSString alloc] initWithFormat:@"--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
                    
                    {
                        NSString *contentDisposition = [[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n\r\n", key];
                        NSData *contentDispositionData = [contentDisposition dataUsingEncoding:NSUTF8StringEncoding];
                        fwrite([contentDispositionData bytes], 1, [contentDispositionData length], file);
                    }
                    //[self appendData:[[[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n\r\n", key] dataUsingEncoding:NSUTF8StringEncoding]];
                    
                    {
                        NSString *objStr = [obj description];
                        NSData *objData = [objStr dataUsingEncoding:NSUTF8StringEncoding];
                        fwrite([objData bytes], 1, [objData length], file);
                    }
                    //[self appendData:[[obj description] dataUsingEncoding:NSUTF8StringEncoding]];
                    
                    {
                        NSString *boundaryStr = [[NSString alloc] initWithFormat:@"\r\n--%@\r\n", boundary];
                        NSData *boundaryData = [boundaryStr dataUsingEncoding:NSUTF8StringEncoding];
                        fwrite([boundaryData bytes], 1, [boundaryData length], file);
                    }
                    //[self appendData:[[[NSString alloc] initWithFormat:@"\r\n--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
                }
            }];
        }
        
        fclose(file);
    }
    
    return filePath;
}

@end

File: iAsync.Swift/lib/JNetwork/JNetwork/HttpBodyBuilder/JHttpBodyBuilder.swift
//
//  JHttpBodyBuilder.swift
//  JNetwork
//
//  Created by Vladimir Gorbenko on 25.09.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

public class JHttpBodyBuilder : NSObject {
    
    public class func tmpFileForUploadStreamWithDataForFilePath(
        dataFilePath: String,
        boundary: String,
        name: String,
        fileName: String,
        dictWithParam: NSDictionary) -> String
    {
        var filePath: String!
        
        autoreleasepool {
            
            filePath = NSUUID().UUIDString
            filePath = NSString.cachesPathByAppendingPathComponent(filePath)
            let filePathPtr = filePath.cStringUsingEncoding(NSUTF8StringEncoding)
            
            let file = fopen(filePathPtr!, "w+")
            
            autoreleasepool {
                
                { () -> () in
                    
                    let boundaryStr  = "--\(boundary)\r\n"
                    let boundaryData = boundaryStr.dataUsingEncoding(NSUTF8StringEncoding)!
                    fwrite(boundaryData.bytes, 1, boundaryData.length, file)
                }();
                //[result appendData:[[[NSString alloc] initWithFormat:@"--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
                    
                { () -> () in
                    
                    let contentDisposition = "Content-Disposition: form-data; name=\"\(name)\"; filename=\"\(fileName)\"\r\n"
                    let contentDispositionData = contentDisposition.dataUsingEncoding(NSUTF8StringEncoding)!
                    fwrite(contentDispositionData.bytes, 1, contentDispositionData.length, file)
                }();
                //[result appendData:[[[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"; filename=\"%@\"\r\n", parameter, fileName] dataUsingEncoding:NSUTF8StringEncoding]];
                
                { () -> () in
                    let contentType     = "Content-Type: application/octet-stream\r\n\r\n";
                    let contentTypeData = contentType.dataUsingEncoding(NSUTF8StringEncoding)!
                    fwrite(contentTypeData.bytes, 1, contentTypeData.length, file)
                }();
                //[result appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
                
                { () -> () in
                    
                    let uploadDataFile = fopen(dataFilePath.cStringUsingEncoding(NSUTF8StringEncoding)!, "r")
                    
                    let bufferLength = 10*1024
                    var array = Array<UInt8>(count: Int(bufferLength), repeatedValue: 0)
                    
                    array.withUnsafeMutableBufferPointer({ (inout cArray: UnsafeMutableBufferPointer<UInt8>) -> () in
                        
                        let readFileChunk = { () -> Int in
                            return fread(cArray.baseAddress, 1, bufferLength, uploadDataFile)
                        }
                        var readBytes = readFileChunk()
                        while readBytes != 0 {
                            
                            fwrite(cArray.baseAddress, 1, readBytes, file)
                            readBytes = readFileChunk()
                        }
                    })
                    
                    fclose(uploadDataFile);
                }();
                //[result appendData:data];
                
                { () -> () in
                    
                    let boundaryStr  = "\r\n--\(boundary)\r\n"
                    let boundaryData = boundaryStr.dataUsingEncoding(NSUTF8StringEncoding)!
                    fwrite(boundaryData.bytes, 1, boundaryData.length, file)
                }()
                //[result appendData:[[[NSString alloc] initWithFormat:@"\r\n--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
            }
            
            { () -> () in
                
                for (key, value) in dictWithParam {
                    
                    autoreleasepool {
                        
                        { () -> () in
                            let boundaryStr  = "--\(boundary)\r\n"
                            let boundaryData = boundaryStr.dataUsingEncoding(NSUTF8StringEncoding)!
                            fwrite(boundaryData.bytes, 1, boundaryData.length, file)
                        }();
                        //[self appendData:[[[NSString alloc] initWithFormat:@"--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
                        
                        { () -> () in
                            let contentDisposition = "Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n"
                            let contentDispositionData = contentDisposition.dataUsingEncoding(NSUTF8StringEncoding)!
                            fwrite(contentDispositionData.bytes, 1, contentDispositionData.length, file)
                        }();
                        //[self appendData:[[[NSString alloc] initWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n\r\n", key] dataUsingEncoding:NSUTF8StringEncoding]];
                            
                        { () -> () in
                            let valueStr  = (value as! NSObject).description
                            let valueData = valueStr.dataUsingEncoding(NSUTF8StringEncoding)!
                            fwrite(valueData.bytes, 1, valueData.length, file)
                        }();
                        //[self appendData:[[obj description] dataUsingEncoding:NSUTF8StringEncoding]];
                        
                        { () -> () in
                            let boundaryStr  = "\r\n--\(boundary)\r\n"
                            let boundaryData = boundaryStr.dataUsingEncoding(NSUTF8StringEncoding)!
                            fwrite(boundaryData.bytes, 1, boundaryData.length, file)
                        }()
                        //[self appendData:[[[NSString alloc] initWithFormat:@"\r\n--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
                    }
                }
            }()
            
            fclose(file)
        }
        
        return filePath
    }
}
